feat(html): added labels to test cases and test files

alexneo2003 opened the pull request:
Related to #14487 

- match labels from test title by @ symbol, e.g. @smoke, @regression
- added labels to test case view
- added labels to test file view
- handling Meta and Ctrl keys for label selection, if no label is selected then add label to search input, if label is selected then remove label
- simple click on label will add it to search input, click another - will change search input to new label
- combining labels with search input and status filters


https://user-images.githubusercontent.com/11705902/223535701-f9971b3d-5eda-4a09-b6af-c1d8bba6ef03.mp4

alexneo2003 committed:
feat(html): added labels to test cases and test files
- match labels from test file title by @ symbol, e.g. @smoke, @regression
- added labels to test case view
- added labels to test file view
- handling Meta and Ctrl keys for label selection, if no label is selected
  then add label to search input, if label is selected then remove label
- simple click on label will add it to search input, click another - will
  change search input to new label
- combining labels with search input and status filters

pavelfeldman commented on packages/html-reporter/src/filter.ts:
```diff
@@ -118,12 +125,16 @@ export class Filter {
         return false;
     }
+    if (this.labels.length) {
+      const matches = this.labels.filter(l => searchValues.text.match(new RegExp(`(\\s|^)${l}(\\s|$)`, 'g'))).length === this.labels.length;
```
`searchValues.text` has just been constructed to contain status, project name, path, title. You don't really want to regex against it. What you probably want it to parse out labels from the test.path + test.title into `searchValues.labels` and check that all the filter labels are present in the `searchValues.labels`.

pavelfeldman commented on packages/html-reporter/src/headerView.tsx:
```diff
@@ -31,12 +31,25 @@ export const HeaderView: React.FC<React.PropsWithChildren<{
-        setFilterText(params.get('q') || '');
-      });
-    })();
+    const popstateFn = () => {
```
Is this necessary? Looks like this code will end up in the Link/Router infrastructure rather than scattered across the components. Let's do one thing at a time.

alexneo2003 replied:
i'm removed inputFn with his listeners but left `popstateFn` with added `removeEventListener` to avoiding duplication listerers

old implementation 

https://user-images.githubusercontent.com/11705902/224674539-39413e69-9184-44d4-aaf2-a596290d00f1.mp4

new

https://user-images.githubusercontent.com/11705902/224674669-18f4e15e-df5e-48f0-b039-a3701fda5a45.mp4

pavelfeldman commented on packages/html-reporter/src/labels.css:
```diff
@@ -0,0 +1,6 @@
+.labels {
```
Missing copyright

pavelfeldman replied:
Let's call this `labelsView` to make it a little less ambiguous.

pavelfeldman commented on packages/html-reporter/src/labels.tsx:
```diff
@@ -0,0 +1,105 @@
+import { navigate } from './links';
+import type { TestCaseSummary } from './types';
+
+/**
```
Misplaced copyright

pavelfeldman commented on packages/html-reporter/src/labels.tsx:
```diff
@@ -0,0 +1,105 @@
+import React from 'react';
```
labelsView as well

pavelfeldman commented on packages/html-reporter/src/labels.tsx:
```diff
@@ -0,0 +1,105 @@
+}>> = ({ testCase, style }) => {
+
+  const labels = React.useMemo(() => {
+    return matchTags(testCase.title).sort((a, b) => a.localeCompare(b));
```
nit: Feels like preprocessing to filter out tags would make code look nicer and more efficient.

pavelfeldman commented on packages/html-reporter/src/labels.tsx:
```diff
@@ -0,0 +1,105 @@
+      if (!q.includes('@')) {
+        searchParams.set('q', `${q} @${tag}`.trim());
+      } else if (!q.includes(`@${tag}`)) {
+        searchParams.set('q', `${q} @${tag}`.trim());
```
Looks same as in line 61, why different branches?

pavelfeldman commented on packages/html-reporter/src/labels.tsx:
```diff
@@ -0,0 +1,105 @@
+      } else if (!q.includes(`@${tag}`)) {
+        searchParams.set('q', `${q} @${tag}`.trim());
+      } else {
+        const re = new RegExp(`@${escapeRegExp(tag)}`, 'g');
```
Any reason to not replace it as a string?

pavelfeldman commented on packages/html-reporter/src/labels.tsx:
```diff
@@ -0,0 +1,105 @@
+    const searchParams = new URLSearchParams(window.location.hash.slice(1));
+    const q = searchParams.get('q') || '';
+
+    if (searchParams.has('testId')){
```
Looks like we have two different types of links: simple links like `ProjectLink` that you use in tests and in this branch and a smart ctrl+clickable link. Let's split them?

pavelfeldman commented on packages/html-reporter/src/testCaseView.tsx:
```diff
@@ -36,7 +37,8 @@ export const TestCaseView: React.FC<{
     {test && <div className='test-case-title'>{test?.title}</div>}
     {test && <div className='test-case-location'>{test.location.file}:{test.location.line}</div>}
-    {test && !!test.projectName && <ProjectLink projectNames={projectNames} projectName={test.projectName}></ProjectLink>}
+    {test && !!test.projectName && <><span style={{ margin: '2px 8px' }}>Project: <ProjectLink projectNames={projectNames} projectName={test.projectName}></ProjectLink></span></>}
```
Looks like you are looking for `.mx-2` style, we use github style here.

pavelfeldman reviewed (commented):
Looks mostly good! Some behavior observations:

- Labels should be rendered similarly, we can remove the "Project" text here and just inline them together:
<img width="185" alt="Screenshot 2023-03-09 at 7 42 50 AM" src="https://user-images.githubusercontent.com/883973/224076611-dedf182c-3115-4344-a364-0211b856ec64.png">
- Clicking back after navigating by a link takes two "Back" clicks. You should use regular <Link> / <Route> where possible to have it just work
- Project and @ labels are rendered inconsistently in list, should all be one line, like project labels:
<img width="627" alt="Screenshot 2023-03-09 at 7 46 14 AM" src="https://user-images.githubusercontent.com/883973/224077322-6b666a8d-ac77-4a50-804d-bae2462a59fb.png">


---

If it helps, I would remove the ctrl+click for now to make the patch smaller and simpler in terms of simply using hrefs. We can always follow-up to add complexity one step at a time.

alexneo2003 commented:
@pavelfeldman hi, i have some question
i'm moving @ labels aside project label and now we have some issue - as you see labels overflowing test title - main part of test title.
screenshot - it's my real project

<img width="1045" alt="Screenshot 2023-03-11 at 16 51 15" src="https://user-images.githubusercontent.com/11705902/224492524-59bb2c4d-6c7b-459c-ad0b-4d848fb67296.png">

i can try do some like this
- for test title using style for wrap text into two lines and ellipsis it
```css
  display: -webkit-box;
  -webkit-line-clamp: 2;
  -webkit-box-orient: vertical;
```
- labels render together with project, if it exists (labels + project zone occupy 40% of row or maybe less)
- if no one - title render for all width

<img width="1009" alt="Screenshot 2023-03-11 at 18 24 16" src="https://user-images.githubusercontent.com/11705902/224495928-657799a4-e4ed-456f-b64e-c3bb1ce82a29.png">

pavelfeldman commented:
I wonder if we should keep it simple and do the GitHub style regular overflow, nothing fancy:

<img width="867" alt="Screenshot 2023-03-11 at 10 57 25 AM" src="https://user-images.githubusercontent.com/883973/224506692-9035484d-dd6a-49a1-b1c3-1a97db9bc0c7.png">

alexneo2003 commented:
@pavelfeldman i have some problem)
i have no idea how but `imageDiffView` tests will failed.
i have no changing nothing at this component but it failed.
report attached.
[html-reporter-ct-test.zip](https://github.com/microsoft/playwright/files/10951368/html-reporter-ct-test.zip)

alexneo2003 committed:
feat(html): add labels to reporter
- search for labels in test path + test title
- testFileView - project link + labels will be displayed inline with test title
- testCaseView - labels will be displayed inline with project link

alexneo2003 commented:
in `labelsView` i'm duplicating from `reportView`
```ts
const testFilesRoutePredicate = (params: URLSearchParams) => !params.has('testId');
const testCaseRoutePredicate = (params: URLSearchParams) => params.has('testId');
```
explain: 
1. in `labelsView` using `Route` and need `testFilesRoutePredicate` functions
2. when i'm run `imageDiffView.spec.tsx` it chaining styles from` ImageDiffView --> TestResultView --> TestCaseView --> LabelsView --> ReportView `(because testFilesRoutePredicate was imported from they) 
3. `ReportView` styles was applied to `ImageDiffView` and style `body { margin: 8px; }` form `user agent style` which was before, will be replaced by a new one --> `body { margin: 0 auto; }`, and three test from `imageDiffView.spec.tsx` will be failed.

we can remove duplication of `testFilesRoutePredicate` and `testCaseRoutePredicate` but then we need to change `imageDiffView.spec.tsx` with new styles.

alexneo2003 commented:
https://user-images.githubusercontent.com/11705902/224677684-bc08172a-ee42-4f02-a5c0-c604235175b8.mp4

alexneo2003 committed:
Merge branch 'main' of https://github.com/alexneo2003/playwright into html-reporter-label

alexneo2003 committed:
test(html): fixing label test

alexneo2003 committed:
Merge branch 'main' of https://github.com/alexneo2003/playwright into html-reporter-label

pavelfeldman commented on packages/html-reporter/src/headerView.tsx:
```diff
@@ -47,7 +50,7 @@ export const HeaderView: React.FC<React.PropsWithChildren<{
         event => {
           event.preventDefault();
-          navigate(`#?q=${filterText ? encodeURIComponent(filterText) : ''}`);
+          navigate(`#?q=${filterText ? decodeURIComponent(filterText) : ''}`);
```
nit: i'm not sure we need to encode/decode it here at all since it is a part of the anchor

alexneo2003 replied:
encodeURIComponent is not correct method to use it
```
   Expected pattern: /@smoke/
    Received string:  "http://localhost:50414/#?q=%40smoke"
```

pavelfeldman replied:
We parse hash as follows: ```const searchParams = new URLSearchParams(window.location.hash.slice(1));```, so if the value of `filterText` is `foo&bar=baz`, you end up with `q==foo`, which is not expected.

pavelfeldman commented on packages/html-reporter/src/labelsView.tsx:
```diff
@@ -0,0 +1,112 @@
+            <LabelClick key={tag} tag={tag} />
+          </Route>
+          <Route predicate={testCaseRoutePredicate}>
+            <LabelLink key={tag} tag={tag} />
```
These are seemingly the same, what is the difference and why do you use route here?

alexneo2003 replied:
There are two labels
 LabelLink - 'simple', inside <a/> and href - used for testCaseView
 LabelClick - 'smart', internal onClick handler that works in conjunction with filters by project or status (status AND label) and is used in testFileView

pavelfeldman replied:
I see. It looks like it is adding unnecessary complexity: labels control should not depend on the route. It would read simpler if you inlined smart and non-smart labels in the code that needs them respectively. Otherwise, you have 2 different clients that need 2 different controls, but you make them go through a single location that makes sense of the context via route and gives away the right control.

pavelfeldman commented on packages/html-reporter/src/labelsView.tsx:
```diff
@@ -0,0 +1,112 @@
+  if (!labels.length)
+    return null;
+
+  const encodedTags = labels.map(tag => encodeTag(tag));
```
What's up with this all encode/decode here? :)

pavelfeldman commented on packages/html-reporter/src/testFileView.css:
```diff
@@ -18,7 +18,7 @@
   align-items: center;
   padding: 2px 10px;
-  white-space: nowrap;
+  white-space: normal;
```
just drop it.

pavelfeldman commented on packages/html-reporter/src/testFileView.css:
```diff
@@ -67,3 +67,7 @@
 }
+
+.test-file-test-status-icon {
+  flex-shrink: 0;
```
nit: we say flex: none of those.

alexneo2003 committed:
fix(html-reporter): work with PR

pavelfeldman commented on packages/html-reporter/src/filter.ts:
```diff
@@ -102,7 +110,8 @@ export class Filter {
         project: test.projectName.toLowerCase(),
-        status: status as any
+        status: status as any,
+        labels: (test.path.join(' ') + test.title).toLowerCase(),
```
At this point, it is not distinguishable from the text in terms of the search, so just use `.text`?

pavelfeldman commented on packages/html-reporter/src/filter.ts:
```diff
@@ -118,12 +127,16 @@ export class Filter {
         return false;
     }
+    if (this.labels.length) {
+      const matches = this.labels.filter(l => searchValues.labels?.match(new RegExp(`(\\s|^)${escapeRegExp(l)}(\\s|$)`, 'g'))).length === this.labels.length;
```
```suggestion
      const matches = this.labels.every(l => searchValues.text?.match(new RegExp(`(\\s|^)${escapeRegExp(l)}(\\s|$)`, 'g')));
```

pavelfeldman commented on packages/html-reporter/src/filter.ts:
```diff
@@ -133,5 +146,6 @@ type SearchValues = {
   text: string;
   project: string;
   status: 'passed' | 'failed' | 'flaky' | 'skipped';
+  labels?: string;
```
Drop?

alexneo2003 committed:
feat(html-reporter): add labels to test file and test case
- separate labels into their own component
- fixint tests

pavelfeldman commented on packages/html-reporter/src/testCaseView.tsx:
```diff
@@ -84,3 +93,19 @@ function retryLabel(index: number) {
+const LabelsLinkView: React.FC<React.PropsWithChildren<{
+  labels: string[],
+}>> = ({ labels }) => {
+  return labels.length > 0 ? (
```
nit: you don't seem to need length check.
